ping/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+mod protocol_builder;
+mod protocol_generated;
+
+use crate::protocol_generated::p2p::ping::*;
+use flatbuffers::{get_root, FlatBufferBuilder};
+use fnv::FnvHashMap;
+use futures::sync::mpsc::Sender;
+use log::debug;
+use p2p::{
+    context::{ServiceContext, SessionContext},
+    traits::{ProtocolMeta, ServiceProtocol},
+    ProtocolId, SessionId,
+};
+use std::{
+    str,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use tokio::codec::length_delimited::LengthDelimitedCodec;
+
+const SEND_PING_TOKEN: u64 = 0;
+const CHECK_TIMEOUT_TOKEN: u64 = 1;
+// TODO: replace this with real PeerId
+type PeerId = SessionId;
+
+/// Ping protocol events
+#[derive(Debug)]
+pub enum Event {
+    /// Peer send ping to us.
+    Ping(PeerId),
+    /// Peer send pong to us.
+    Pong(PeerId, Duration),
+    /// Peer is timeout.
+    Timeout(PeerId),
+    /// Peer cause a unexpected error.
+    UnexpectedError(PeerId),
+}
+
+pub struct PingProtocol {
+    id: ProtocolId,
+    /// the interval that we send ping to peers.
+    interval: Duration,
+    /// consider peer is timeout if during a timeout we still have not received pong from a peer
+    timeout: Duration,
+    event_sender: Sender<Event>,
```

Comment:
Depending on the context, and the example, I think there might not need a `futures sender` here(maybe a crossbeam channel?), or consider implementing a `newtype` structure?

Reply:
Why?